### PR TITLE
Properly compute resource folder when linking statically

### PR DESCRIPTION
**Pull request by drexin**
- deduplicate the logic to compute the resource folder
- install headers and module files in shared and static resource folders
- forward -static flag when calling swiftc with -print-target-info

### CODE REVIEW

**Comment by drexin** [comment]
@swift-ci test

File: include/swift/Driver/ToolChain.h [hunk 1]
```diff
@@ -205,7 +205,7 @@ class ToolChain {
 
   /// Get the resource dir link path, which is platform-specific and found
   /// relative to the compiler.
-  void getResourceDirPath(SmallVectorImpl<char> &runtimeLibPath,
+  void getResourceDirPath(SmallString<128> &runtimeLibPath,
```
Comment (compnerd) [inline include/swift/Driver/ToolChain.h:208]: Why this change?  The `SmallVectorImpl` is the stack allocation size erased base type for `SmallString`.  This allows for different sizes to be used.

File: include/swift/Frontend/Frontend.h [hunk 2]
```diff
@@ -128,7 +128,8 @@ class CompilerInvocation {
   bool parseArgs(ArrayRef<const char *> Args, DiagnosticEngine &Diags,
                  SmallVectorImpl<std::unique_ptr<llvm::MemoryBuffer>>
                      *ConfigurationFileBuffers = nullptr,
-                 StringRef workingDirectory = {});
+                 StringRef workingDirectory = {},
+                 StringRef mainExecutablePath = {});
```
Comment (compnerd) [inline include/swift/Frontend/Frontend.h:132]: Indentation seems off

File: include/swift/Frontend/Frontend.h [hunk 3]
```diff
@@ -214,8 +215,16 @@ class CompilerInvocation {
   /// executable.
   static void computeRuntimeResourcePathFromExecutablePath(
       StringRef mainExecutablePath,
+      bool shared,
       llvm::SmallString<128> &runtimeResourcePath);
```
Comment (compnerd) [inline include/swift/Frontend/Frontend.h:219]: `clang-format` please

File: lib/Driver/Driver.cpp [hunk 4]
```diff
@@ -2255,6 +2255,14 @@ bool Driver::handleImmediateArgs(const ArgList &Args, const ToolChain &TC) {
       commandLine.push_back(resourceDirArg->getValue());
     }
 
+    if (Args.hasFlag(options::OPT_static_executable,
+                     options::OPT_no_static_executable, false) ||
+        Args.hasFlag(options::OPT_static_stdlib,
+                     options::OPT_no_static_stdlib, false) ||
+        Args.hasArg(options::OPT_static)) {
+      commandLine.push_back("-static");
+    }
```
Comment (compnerd) [inline lib/Driver/Driver.cpp:2264]: I thought `-static-stdlib` was for statically linking the standard library but creating a dynamic executable/library.  Is that not the case?  If not, this makes sense, otherwise, this seems wrong because `-static` will will always do static linking.

**Review by compnerd — COMMENTED** [review]
Could you please add a test for the driver path computation?

File: lib/Driver/Driver.cpp (hunk 4 above)
Comment (drexin) [reply lib/Driver/Driver.cpp:2264]: The linking is handled by the driver and `OPT_static` is not used in any form in the frontend currently. I'm happy to change it to something else, but the frontend does not seem to know `OPT_static_stdlib` or `OPT_static_executable`.

File: include/swift/Driver/ToolChain.h (hunk 1 above)
Comment (drexin) [reply include/swift/Driver/ToolChain.h]: reverted

File: include/swift/Frontend/Frontend.h [hunk 5]
```diff
@@ -213,9 +214,14 @@ class CompilerInvocation {
   /// Computes the runtime resource path relative to the given Swift
   /// executable.
   static void computeRuntimeResourcePathFromExecutablePath(
-      StringRef mainExecutablePath,
+      StringRef mainExecutablePath, bool shared,
       llvm::SmallString<128> &runtimeResourcePath);
 
+  /// Appends `lib/swift[_static]` to the given path
+  static void appendLibDir(llvm::SmallString<128> &path, bool shared);
```
Comment (compnerd) [inline include/swift/Frontend/Frontend.h:221]: I would say that `SmallVectorImpl<char>` is better than the hardcoded stack size.

File: include/swift/Frontend/Frontend.h [hunk 6]
```diff
@@ -213,9 +214,14 @@ class CompilerInvocation {
...
+
+  void computeRuntimeResourcePath(llvm::SmallString<128> &runtimeResourcePath);
```
Comment (compnerd) [inline include/swift/Frontend/Frontend.h:223]: Similar

File: lib/Driver/Driver.cpp (hunk 4 above)
Comment (compnerd) [reply lib/Driver/Driver.cpp:2264]: Hmm, I suppose that my discomfort comes from the difference in what `-static` means when you are using `clang` and `swift`.  It just seems very easy to confuse someone (as opposed to many of the other arguments that have different spellings).

`-static`, `-static-executable` both imply creating a static binary.  `-static-stdlib` says that statically link the standard library but still still generate a dynamically linked binary.  As long `-static` is not forwarded from the frontend to the driver it will be okay I suppose, just confusing (non-Darwin targets use the linker driver (`clang`) to invoke the linker which will process `-static` as generating a static binary).  If you want to still go with this, I think that this deserves a comment explaining what is going on here and how this fits together.

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test macos

**Comment by drexin** [comment]
@swift-ci test macos

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test windows

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test windows

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test macos

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test macos

**Comment by drexin** [comment]
@swift-ci test windows

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test windows

**Comment by drexin** [comment]
@swift-ci test macos

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test linux

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test

**Comment by drexin** [comment]
@swift-ci test windows

**Comment by drexin** [comment]
@swift-ci test windows

File: lib/Driver/Driver.cpp (hunk 4 above)
Comment (drexin) [reply lib/Driver/Driver.cpp:2263]: I added a new flag to the frontend `-use-static-resource-dir`

File: include/swift/Frontend/Frontend.h [hunk 7]
```diff
@@ -213,8 +214,11 @@ class CompilerInvocation {
   /// Computes the runtime resource path relative to the given Swift
   /// executable.
   static void computeRuntimeResourcePathFromExecutablePath(
-      StringRef mainExecutablePath,
-      llvm::SmallString<128> &runtimeResourcePath);
+      StringRef mainExecutablePath, bool shared,
+      llvm::SmallVectorImpl<char> &runtimeResourcePath);
+
+  /// Appends `lib/swift[_static]` to the given path
+  static void appendLibDir(llvm::SmallVectorImpl<char> &path, bool shared);
```
Comment (compnerd) [inline include/swift/Frontend/Frontend.h:221]: Would be nice to make this `appendSwiftLibDir` since `libdir` in most contexts here tends to refer to `lib` vs `lib32` or `lib64`.

File: include/swift/Frontend/FrontendOptions.h [hunk 8]
```diff
@@ -283,6 +283,8 @@ class FrontendOptions {
   /// -dump-scope-maps.
   SmallVector<std::pair<unsigned, unsigned>, 2> DumpScopeMapLocations;
 
+  bool UseSharedResourceFolder = true;
```
Comment (compnerd) [inline include/swift/Frontend/FrontendOptions.h:286]: Doc comment please (its obvious in the context of the change, but less so when you just have a type to work with).

**Review by compnerd — APPROVED** [review]

**Comment by drexin** [comment]
@swift-ci smoke test

**Comment by drexin** [comment]
@swift-ci smoke test

**Comment by drexin** [comment]
@swift-ci smoke test windows

**Comment by drexin** [comment]
@swift-ci smoke test windows

File: lib/Driver/ToolChains.cpp [hunk 9]
```diff
@@ -1266,24 +1274,18 @@ void ToolChain::getClangLibraryPath(const ArgList &Args,
...
-  // need to be the same.
   if (const Arg *A = args.getLastArg(options::OPT_resource_dir)) {
     StringRef value = A->getValue();
     resourceDirPath.append(value.begin(), value.end());
   } else if (!getTriple().isOSDarwin() && args.hasArg(options::OPT_sdk)) {
     StringRef value = args.getLastArg(options::OPT_sdk)->getValue();
     resourceDirPath.append(value.begin(), value.end());
-    llvm::sys::path::append(resourceDirPath, "usr", "lib",
-                            shared ? "swift" : "swift_static");
+    llvm::sys::path::append(resourceDirPath, "usr");
```
Comment (DougGregor) [inline lib/Driver/ToolChains.cpp:1283]: Why are we adding `<resource-dir>/usr` now?

Comment (drexin) [reply lib/Driver/ToolChains.cpp:1283]: We are using the SDK root here as opposed to the path relative to the executable. In the executable case we get something like `$PATH/bin/swift` and remove the `bin/swiftc` part from it and append `lib/swift[_static]`. The SDK path starts one level above that, so we have to add `usr` only in this case.

**Review by DougGregor — APPROVED** [review]
